pilopt/src/lib.rs

```diff
@@ -121,33 +122,41 @@ fn remove_unreferenced_definitions<T: FieldElement>(pil_file: &mut Analyzed<T>)
     pil_file.remove_trait_impls(&impls_to_remove);
 }
 
-/// Builds a lookup-table that can be used to turn array elements
-/// (in form of their poly ids) into the names of the arrays.
+/// Builds a lookup-table that can be used to turn all poly ids into the names of the symbols that define them.
+/// For array elements, this contains the array name and the index of the element in the array.
 fn build_poly_id_to_definition_name_lookup(
     pil_file: &Analyzed<impl FieldElement>,
-) -> BTreeMap<PolyID, &String> {
+) -> BTreeMap<PolyID, (&String, Option<usize>)> {
     let mut poly_id_to_definition_name = BTreeMap::new();
     #[allow(clippy::iter_over_hash_type)]
     for (name, (symbol, _)) in &pil_file.definitions {
         if matches!(symbol.kind, SymbolKind::Poly(_)) {
-            symbol.array_elements().for_each(|(_, id)| {
-                poly_id_to_definition_name.insert(id, name);
-            });
+            symbol
+                .array_elements()
+                .enumerate()
+                .for_each(|(idx, (_, id))| {
+                    let array_pos = symbol.is_array().then_some(idx);
+                    poly_id_to_definition_name.insert(id, (name, array_pos));
+                });
         }
     }
     #[allow(clippy::iter_over_hash_type)]
     for (name, (symbol, _)) in &pil_file.intermediate_columns {
-        symbol.array_elements().for_each(|(_, id)| {
-            poly_id_to_definition_name.insert(id, name);
-        });
+        symbol
+            .array_elements()
+            .enumerate()
+            .for_each(|(idx, (_, id))| {
+                let array_pos = symbol.is_array().then_some(idx);
+                poly_id_to_definition_name.insert(id, (name, array_pos));
+            });
     }
     poly_id_to_definition_name
```

Comment:
If this is already changing, we could turn it into something like

```rust
pil_file.definitions.iter().filter(|d| matches!(symbol.kind, SymbolKind::Poly(_))).chain(&pil_file.intermediate_columns).flat_map(|(name, (symbol, _))| symbol
            .array_elements()
            .enumerate()
            .map(|(idx, (_, id))| {
                let array_pos = symbol.is_array().then_some(idx);
                (id, (name, array_pos))
            })
   ).collect()
   ```
   
   to avoid side effects and code repetition